File: Framework/World/Entities/Tree.cs
namespace Biosphere {
    public class Tree : Entity {

        public int growthState;
        private int clock = 0;
        private static string[] states = ["sprout", "sapling", "mature", "ancient", "decaying"];
        private static int[] time = [50, 75, 300, 500, 700];

        public override string Type => "Tree";

        public Tree() : base() {
            growthState = 0;
            state = states[0];
        }

        protected override void Update() {
            if (World.GetTile(Pos).HasState("fire")) {
                if (growthState < 3) {
                    growthState = 3;
                    clock = 0;
                }
                clock += 10;
            }

            if (growthState > 4) {
                World.RemoveEntity(this);
            } else {
                state = states[growthState];
            }

            ++clock;
            if (clock >= time[growthState]) {
                ++growthState;
                clock = 0;
            }
        }
    }
}

File: Viewer/src/framework/entities/tree.ts
import { LoadImage, Sprite } from "../../lib/renderer.js";
import { World } from "../../lib/world.js";
import { Entity } from "../entity.js";

const ancient = LoadImage("./images/entities/tree.png");

const states = new Map();
states.set("sprout", LoadImage("./images/entities/sprout.png"));
states.set("sapling", LoadImage("./images/entities/sapling.png"));
states.set("mature", LoadImage("./images/entities/mature.png"));
states.set("ancient", LoadImage("./images/entities/ancient.png"));
states.set("decaying", LoadImage("./images/entities/decayed.png"));

export class Tree extends Entity {
    sprite: Sprite;

    constructor(world: World) {
        super(world);

        this.sprite = new Sprite(ancient, world.tileSize, world.tileSize * 1.5);
        this.sprite.layer = 0;
        this.sprite.offset.y = -this.world.tileSize * 0.25;

        this.update();
    }

    public update() {
        this.sprite.image = states.get(this.state);
        if (this.sprite.image === undefined) this.world.renderer.removeSprite(this.sprite);
        else this.world.renderer.addSprite(this.sprite);
        this.sprite.pos.x = this.pos.x * this.world.tileSize * 0.5; 
        this.sprite.pos.y = this.pos.y * this.world.tileSize * 0.5;
    }

    public destroy(): void {
        this.world.renderer.removeSprite(this.sprite);
    }
}